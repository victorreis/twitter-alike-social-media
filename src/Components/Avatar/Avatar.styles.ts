import styled, { CSSObject } from 'styled-components';

import { toPx } from '../../Utils/Transform/toPx.util';
import { Typography } from '../Typography';
import {
  AvatarContainerStyleProps,
  AvatarSize,
  AvatarSizes,
  AvatarStyleProps,
} from './Avatar.types';

export const avatarDimensions: AvatarSizes = {
  SM: 32,
  MD: 56,
  LG: 100,
};

const getAvatarDimensions = ({ size }: { size: AvatarSize }): CSSObject => {
  return {
    width: avatarDimensions[size],
    height: avatarDimensions[size],
  };
};

const getAvatarContainerDimensions = ({
  size,
}: {
  size: AvatarSize;
}): CSSObject => {
  return {
    width: avatarDimensions[size] + 4,
    height: avatarDimensions[size] + 4,
  };
};

export const AvatarContainer = styled.div<AvatarContainerStyleProps>`
  vertical-align: middle;
  border-radius: 50%;
  background-color: white;
  display: grid;
  place-items: center;
  box-sizing: border-box;
  padding: 2px;
  cursor: ${({ isAvatarClickable }) =>
    isAvatarClickable ? 'pointer' : undefined};

  ${getAvatarContainerDimensions};
`;

export const AvatarImage = styled.img<AvatarStyleProps>`
  vertical-align: middle;
  border-radius: 50%;
  ${getAvatarDimensions}
`;

export const AvatarInitials = styled(Typography)<AvatarStyleProps>`
  font-size: ${({ size }: { size: AvatarSize }) =>
    toPx(avatarDimensions[size] * 0.67)};
  color: red;
  font-family: Impact;
`;
