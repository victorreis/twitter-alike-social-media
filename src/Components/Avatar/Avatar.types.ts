import { TestProps } from '../../Config/Tests/Test.types';
import { Size } from '../../Theme/Types/Sizes.types';

export const availableAvatarSize: ReadonlyArray<
  Extract<Size, 'SM' | 'MD' | 'LG'>
> = ['SM', 'MD', 'LG'] as const;
export type AvatarSize = typeof availableAvatarSize[number];
export type AvatarSizes = Record<AvatarSize, number>;

export interface AvatarClickProps {
  /**
   * Controls when the Avatar is clickable or not.
   */
  isAvatarClickable?: boolean;

  /**
   * Callback function that is called when the Avatar is clicked.
   */
  onClickAvatar?: (nickname: string) => void;
}

export interface RequiredAvatarProps {
  /**
   * Avatar image.
   */
  thumbnailUrl: string;

  /**
   * Name show as title or as initials.
   */
  name: string;

  /**
   * Name show as title or as initials.
   */
  nickname: string;
}

export interface DefaultAvatarProps {
  /**
   * Sets the width and height Avatar dimensions.
   * @default 'MD'.
   */
  size?: AvatarSize;
}

export interface OptionalAvatarProps {}

export type AvatarProps = RequiredAvatarProps &
  DefaultAvatarProps &
  OptionalAvatarProps &
  AvatarClickProps &
  TestProps &
  Omit<React.HTMLAttributes<HTMLImageElement>, 'children' | 'onClick'>;

export type AvatarContainerStyleProps = Required<
  DefaultAvatarProps & Pick<AvatarClickProps, 'isAvatarClickable'>
>;
export type AvatarStyleProps = Required<DefaultAvatarProps>;
