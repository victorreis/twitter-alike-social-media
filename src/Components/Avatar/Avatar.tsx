import { useState } from 'react';

import { TestProps } from '../../Config/Tests/Test.types';
import { AvatarContainer, AvatarImage, AvatarInitials } from './Avatar.styles';
import { AvatarProps, DefaultAvatarProps } from './Avatar.types';

export const avatarDefaults: Required<DefaultAvatarProps> &
  Required<TestProps> = {
  testID: 'Avatar',
  size: 'MD',
};

export const Avatar: React.FC<AvatarProps> = (props): JSX.Element => {
  const {
    testID = avatarDefaults.testID,
    size = avatarDefaults.size,
    thumbnailUrl,
    name,
    nickname,
    isAvatarClickable,
    onClickAvatar,
    style,
    ...others
  } = props;

  const [imageLoadingError, setImageLoadingError] = useState<boolean>(false);

  const handleImageLoadingError = () => {
    setImageLoadingError(true);
  };

  const handleClick = () => {
    if (onClickAvatar) {
      onClickAvatar(nickname);
    }
  };

  const extractInitials = () => {
    const splittedName = name.split(' ');
    let initials = splittedName[0]?.charAt(0) || '';
    if (splittedName.length > 1) {
      initials += splittedName[1]?.charAt(0) || '';
    }
    return initials;
  };

  const description = `${name}'s avatar photo.`;

  const renderImage = () => {
    if (!imageLoadingError) {
      return (
        <AvatarImage
          alt={description}
          data-testid={`${testID}_image`}
          onError={handleImageLoadingError}
          size={size}
          src={thumbnailUrl}
          title={description}
          {...others}
        />
      );
    }
    return null;
  };

  const renderInitials = () => {
    const initials = extractInitials();
    if (imageLoadingError) {
      return (
        <AvatarInitials
          data-testid={`${testID}_initials`}
          size={size}
          variant="body1"
          {...others}
        >
          {initials}
        </AvatarInitials>
      );
    }
    return null;
  };

  return (
    <AvatarContainer
      data-testid={testID}
      isAvatarClickable={Boolean(isAvatarClickable)}
      onClick={handleClick}
      size={size}
      style={style}
    >
      {renderImage()}
      {renderInitials()}
    </AvatarContainer>
  );
};
