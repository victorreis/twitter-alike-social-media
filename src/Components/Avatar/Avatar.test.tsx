import {
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { toPx } from '../../Utils/Transform/toPx.util';
import { Avatar, avatarDefaults } from './Avatar';
import { avatarDimensions } from './Avatar.styles';
import { RequiredAvatarProps, AvatarProps } from './Avatar.types';

describe('avatar component tests', () => {
  const thumbnailUrl =
    'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACgAAAAcCAYAAAATFf3WAAAAAXNSR0IArs4c6QAAArJJREFUWEftlVtIk2EYx3/bPKERZOkItUSIskLojKDRwRrDIlMo0wkzwwmpQzS7KPBQ1E3zstZuUjONsIvAxNKMJPDCykM1yQ5WHppmIeWhMl1839xyUG4xJS/28V297/O+///3ew6fxGw2m1nAj8Rt0MXsuAm6CBA3QTdBVwm4et5BDVpnuMROx4z9bJfgaF84bh9juVC450/rv+UcNsn4+BdytLn8nLI3lZauoa2lmcet7TaRxCQVu3dEYTAYuF1bh6+vH5laLX1vX1J/74FNVbgpOCSUgoJTSF0xKJD6Pj5CYeEZRkdHuHRRzx5lLOvC13BElcTpk7l0ve7nYNw+UfxAfAKDPW84lKzmRH4+TE6wyD+AiNVhND1sprGhnve9A6jVKpYHhZCTk4UU6axVMCtBCzOz+A5/MiGXB6MvvUZqymEkTKJU7AXPJdTWVIschN/6+eICis7paH3aQfiqMIu4xJLG7Ix06pseYTQ+EeabSF4yvfc3lw5TbD04PGQiUB7EZdFgIjCFUhFDa8cLtm3djFTmRenVMvq6u4iO3sm3HxMkq1IoPluEPGCZaCYrI52GphaMxjYH3P6hBh0ZNH0cIy8vG6lURnxCPF4eMnp7eyjR6dDrDWyJ3M79xjvIJJA5bbDT2Oag8ubQIJ7+YorFYjBPMT42iq+fn1gW2uMayipvMTDUh4+Hx/8x2P7sFdFRkaLBpBQ1NTcrGfz8ldCVK7heVcUuRSyVFVfEfWuK54Gg0M1jXNCVsD8ugYj1a0XB6htVPO/ssuUjRqEkcOliyssr+GAaYMPGTaQdTcXH2xthItytq6X7XT8azTGhPZya4U43iWWoii05Y56Z7YSsA1zsaFvkTCOWVSFuHgw69cFzHvQPBOdc26kL3QadwjRLkJugm6CrBFw9v+Br8BdSpWm8K4SiiQAAAABJRU5ErkJggg==';
  const name = 'Victor Reis';
  const nickname = 'victorreis';
  const newSize = 'LG';

  const requiredProps: RequiredAvatarProps = {
    thumbnailUrl,
    name,
    nickname,
  };

  const setup = (props?: AvatarProps) => {
    const renderRTR = () =>
      renderRTRCreator<AvatarProps>(Avatar, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<AvatarProps>(Avatar, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe('behavior tests', () => {
    it(`should render the component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(avatarDefaults.testID);

      expect(testInstance).toBeTruthy();
    });

    it(`should render the image`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const element = screen.getByTestId(avatarDefaults.testID);

      expect(element).toBeInTheDocument();
    });

    it(`should render '${avatarDefaults.size}' as the default 'size'`, () => {
      expect.assertions(1);
      const instance = setup().renderRTR().root;
      const element = instance.findByProps({
        size: avatarDefaults.size,
      });

      expect(element).toBeTruthy();
    });

    it(`should override the default 'size' when it is passed as prop`, () => {
      expect.assertions(1);
      const instance = setup({
        ...requiredProps,
        size: newSize,
      }).renderRTR().root;
      const element = instance.findByProps({ size: newSize });

      expect(element).toBeTruthy();
    });
  });

  describe('style tests', () => {
    it(`should have styled the Container component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const container = screen.getByTestId(avatarDefaults.testID);

      expect(container).toHaveStyle({
        width: toPx(avatarDimensions[avatarDefaults.size] + 4),
        height: toPx(avatarDimensions[avatarDefaults.size] + 4),
      });
    });
  });

  describe('snapshot tests', () => {
    it(`should render correctly`, () => {
      expect.assertions(1);
      const generatedJson = setup().renderRTR().toJSON();

      expect(generatedJson).toMatchSnapshot();
    });
  });
});
