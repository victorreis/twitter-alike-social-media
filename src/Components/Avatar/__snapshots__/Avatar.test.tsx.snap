// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`avatar component tests snapshot tests should render correctly 1`] = `
<div
  className="AvatarContainer-twitter-alike-social-media__sc-ne4tqs-0 dwSceS"
  data-testid="Avatar"
  onClick={[Function]}
  size="MD"
>
  <img
    alt="Victor Reis's avatar photo."
    className="AvatarImage-twitter-alike-social-media__sc-ne4tqs-1 cdAMXQ"
    data-testid="Avatar_image"
    onError={[Function]}
    size="MD"
    src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACgAAAAcCAYAAAATFf3WAAAAAXNSR0IArs4c6QAAArJJREFUWEftlVtIk2EYx3/bPKERZOkItUSIskLojKDRwRrDIlMo0wkzwwmpQzS7KPBQ1E3zstZuUjONsIvAxNKMJPDCykM1yQ5WHppmIeWhMl1839xyUG4xJS/28V297/O+///3ew6fxGw2m1nAj8Rt0MXsuAm6CBA3QTdBVwm4et5BDVpnuMROx4z9bJfgaF84bh9juVC450/rv+UcNsn4+BdytLn8nLI3lZauoa2lmcet7TaRxCQVu3dEYTAYuF1bh6+vH5laLX1vX1J/74FNVbgpOCSUgoJTSF0xKJD6Pj5CYeEZRkdHuHRRzx5lLOvC13BElcTpk7l0ve7nYNw+UfxAfAKDPW84lKzmRH4+TE6wyD+AiNVhND1sprGhnve9A6jVKpYHhZCTk4UU6axVMCtBCzOz+A5/MiGXB6MvvUZqymEkTKJU7AXPJdTWVIschN/6+eICis7paH3aQfiqMIu4xJLG7Ix06pseYTQ+EeabSF4yvfc3lw5TbD04PGQiUB7EZdFgIjCFUhFDa8cLtm3djFTmRenVMvq6u4iO3sm3HxMkq1IoPluEPGCZaCYrI52GphaMxjYH3P6hBh0ZNH0cIy8vG6lURnxCPF4eMnp7eyjR6dDrDWyJ3M79xjvIJJA5bbDT2Oag8ubQIJ7+YorFYjBPMT42iq+fn1gW2uMayipvMTDUh4+Hx/8x2P7sFdFRkaLBpBQ1NTcrGfz8ldCVK7heVcUuRSyVFVfEfWuK54Gg0M1jXNCVsD8ugYj1a0XB6htVPO/ssuUjRqEkcOliyssr+GAaYMPGTaQdTcXH2xthItytq6X7XT8azTGhPZya4U43iWWoii05Y56Z7YSsA1zsaFvkTCOWVSFuHgw69cFzHvQPBOdc26kL3QadwjRLkJugm6CrBFw9v+Br8BdSpWm8K4SiiQAAAABJRU5ErkJggg=="
    title="Victor Reis's avatar photo."
  />
</div>
`;
